easybuild/tools/filetools.py

```diff
@@ -407,26 +407,41 @@ def remove(paths):
             raise EasyBuildError("Specified path to remove is not an existing file or directory: %s", path)
 
 
+def get_cwd(must_exist=True):
+    """
+    Retrieve current working directory
+    """
+    try:
+        cwd = os.getcwd()
+    except FileNotFoundError as err:
+        if must_exist is True:
+            raise EasyBuildError("Working directory does not exist")
+        else:
+            _log.debug("Failed to determine current working directory, but proceeding anyway: %s", err)
+            cwd = None
+
+    return cwd
+
+
 def change_dir(path):
     """
     Change to directory at specified location.
 
     :param path: location to change to
     :return: previous location we were in
     """
-    # determining the current working directory can fail if we're in a non-existing directory
-    try:
-        cwd = os.getcwd()
-    except OSError as err:
-        _log.debug("Failed to determine current working directory (but proceeding anyway: %s", err)
-        cwd = None
+    # determine origin working directory: can fail if non-existent
+    prev_dir = get_cwd(must_exist=False)
 
     try:
         os.chdir(path)
     except OSError as err:
-        raise EasyBuildError("Failed to change from %s to %s: %s", cwd, path, err)
+        raise EasyBuildError("Failed to change from %s to %s: %s", prev_dir, path, err)
 
-    return cwd
+    # determine final working directory: must exist
+    get_cwd()
```

Comment:
I don't see the point of this, if the directory doesn't exist then `os.chdir` above would've failed?

Reply:
This is meant to catch filesystems in a faulty state, where some things work, like listing directories, but still have an undefined `CWD`. It's a rarity, but we experience such situations with the NFS mount of one of our storage systems, which can end up in a very weird state after complex builds that involve creation of many files and folders and jumping around a lot (for instance the issue with OpenFOAM in https://github.com/easybuilders/easybuild-framework/pull/4390)

Reply:
I added an extra comment clarifying the purpose of this line